src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/ConnectionLifetime.cs

```diff
@@ -0,0 +1,157 @@
+            {
+                _frame.Start();
+                _frameStartedTcs.SetResult(null);
+            }
+            else
+            {
+                // Ensure that IConnectionAdapter.OnConnectionAsync does not run on the transport thread.
+                _context.ServiceContext.ThreadPool.UnsafeRun(state =>
+                {
+                    // ApplyConnectionAdaptersAsync should never throw. If it succeeds, it will call _frame.Start().
+                    // Otherwise, it will close the connection.
+                    var ignore = ((ConnectionLifetime)state).ApplyConnectionAdaptersAsync();
+                }, this);
+            }
+        }
+
+        public async Task StopAsync()
+        {
+            await _frameStartedTcs.Task;
+            await _frame.StopAsync();
+            await _adaptedPipelineTask;
+        }
+
+        public void Abort(Exception ex)
+        {
+            _frame.Abort(ex);
+        }
+
+        public void SetBadRequestState(RequestRejectionReason reason)
+        {
+            _frame.SetBadRequestState(reason);
+        }
+
+        private async Task ApplyConnectionAdaptersAsync()
+        {
+            try
+            {
+                var rawSocketOutput = _frame.Output;
+                var rawStream = new RawStream(_frame.Input, rawSocketOutput);
+                var adapterContext = new ConnectionAdapterContext(rawStream);
+                var adaptedConnections = new IAdaptedConnection[_connectionAdapters.Count];
+
+                for (var i = 0; i < _connectionAdapters.Count; i++)
+                {
+                    var adaptedConnection = await _connectionAdapters[i].OnConnectionAsync(adapterContext);
+                    adaptedConnections[i] = adaptedConnection;
+                    adapterContext = new ConnectionAdapterContext(adaptedConnection.ConnectionStream);
+                }
+
+                if (adapterContext.ConnectionStream != rawStream)
+                {
+                    _filteredStream = adapterContext.ConnectionStream;
+                    _adaptedPipeline = new AdaptedPipeline(
+                        adapterContext.ConnectionStream,
+                        PipeFactory.Create(AdaptedPipeOptions),
+                        PipeFactory.Create(AdaptedPipeOptions));
+
+                    _frame.Input = _adaptedPipeline.Input.Reader;
+                    _frame.Output = _adaptedPipeline.Output;
+
+                    _adaptedPipelineTask = RunAdaptedPipeline();
+                }
+
+                _frame.AdaptedConnections = adaptedConnections;
+                _frame.Start();
+                _frameStartedTcs.SetResult(null);
+            }
+            catch (Exception ex)
+            {
+                Log.LogError(0, ex, $"Uncaught exception from the {nameof(IConnectionAdapter.OnConnectionAsync)} method of an {nameof(IConnectionAdapter)}.");
+                _frameStartedTcs.SetResult(null);
+                CloseRawPipes();
```

Comment:
I doesn't. This isn't the consuming class. The closing of the pipes was different for adapters and raw pipes prior to this PR. I just kept existing behavior. We can clean up in another PR if desired.